sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -346,14 +430,167 @@ pi_result piDeviceGetInfo(pi_device device, pi_device_info paramName,
+    if (!isPVC(device)) {
+      cl_int result =
+          clGetDeviceInfo(getClDevice(device), cast<cl_device_info>(paramName),
+                          paramValueSize, paramValue, paramValueSizeRet);
+      return static_cast<pi_result>(result);
+    } else {
+      // For root-device there is no partitioning to report.
+      if (isRootDevice(device))
+        return return_value(pi_device_partition_property{0});
+      if (!isCCS(device)) { // is subdevice
+        struct {
+          pi_device_partition_property arr[3];
+        } partition_properties = {{PI_DEVICE_PARTITION_BY_AFFINITY_DOMAIN,
+                                   PI_DEVICE_AFFINITY_DOMAIN_NEXT_PARTITIONABLE,
+                                   0}};
+        return return_value(partition_properties);
+      } else { // it is CCS
+        struct {
+          pi_device_partition_property arr[2];
+        } partition_properties = {{PI_EXT_INTEL_DEVICE_PARTITION_BY_CSLICE, 0}};
+        return return_value(partition_properties);
+      }
+    }
+    return return_value(pi_device_partition_property{0});
+  }
+
   default:
-    cl_int result = clGetDeviceInfo(
-        cast<cl_device_id>(device), cast<cl_device_info>(paramName),
-        paramValueSize, paramValue, paramValueSizeRet);
+    cl_int result =
+        clGetDeviceInfo(getClDevice(device), cast<cl_device_info>(paramName),
+                        paramValueSize, paramValue, paramValueSizeRet);
     return static_cast<pi_result>(result);
   }
 }
 
+pi_result piDevicePartition(pi_device device,
+                            const pi_device_partition_property *properties,
+                            pi_uint32 num_devices, pi_device *out_devices,
+                            pi_uint32 *out_num_devices) {
+  cl_int result = CL_DEVICE_NOT_FOUND;
+  if (isRootDevice(device)) {
+    result = clCreateSubDevices(
+        getClDevice(device),
+        cast<const cl_device_partition_property *>(properties),
+        cast<cl_uint>(num_devices), cast<cl_device_id *>(out_devices),
+        out_num_devices);
+  } else if (!isCCS(device)) {
+    cl_queue_family_properties_intel qfprops[3];
+    size_t qsize = 0;
+    pi_uint32 family = 0;
+    cl_uint sub_device_count = 0;
+    clGetDeviceInfo(getClDevice(device),
+                    CL_DEVICE_QUEUE_FAMILY_PROPERTIES_INTEL, sizeof(qfprops),
+                    qfprops, &qsize);
+    qsize = qsize / sizeof(cl_queue_family_properties_intel);
+    for (size_t q = 0; q < qsize; q++) {
+      if (qfprops[q].capabilities == CL_QUEUE_DEFAULT_CAPABILITIES_INTEL &&
+          qfprops[q].count > sub_device_count) {
+        family = q;
+        sub_device_count = qfprops[q].count;
+      }
+    }
+    *out_num_devices = sub_device_count;
+    if (!out_devices)
+      return PI_SUCCESS;
+    for (uint32_t i = 0; i < *out_num_devices; ++i) {
+      out_devices[i] = cast<pi_device>(new cl_device_id());
+      csliceSubDevInfo info;
+      info.cl_dev = cast<cl_device_id>(device);
```

Comment:
Don't we need to retain original OpenCL device here?

Reply:
Agreed. I added the code to retain the CL device in Line 574. Thanks